### PR TITLE
Fix graceful shutdown (#363)

### LINKED ISSUE #363: core module class `GracefulShutdownHandler` import wrong type of  HttpRequest
core module `GracefulShutdownHandler` import wrong type of  HttpRequest, cause method
 `boolean isRequestStart(Object msg) {
        return msg instanceof HttpRequest;
    }` useless. It should be `com.linecorp.armeria.common.http.HttpRequest` rather than `io.netty.handler.codec.http.HttpRequest`.

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

Graceful shutdown has been broken since the streaming rewrite. It only
worked for HTTP/1 traffic.

Modifications:

- Rename GracefulShutdownHandler to GracefulShutdownSupport
  - Make it an abstract class and provide two implementations; the
    default one and the disabled one
  - GracefulShutdownSupport is not a ChannelHandler anymore but just a
    utility class
- HttpServerHandler now requires GracefulShutdownSupport to count the
  number of pending responses.
- Add more tests and improve the existing ones in
  GracefulShutdownIntegrationTest

Result:

- Graceful shutdown should work as expected.
- Fixes #363

### CODE REVIEW

**Comment by trustin** [comment]
@anuraaga @jonefeewang PTAL

File: core/src/main/java/com/linecorp/armeria/server/http/HttpServerHandler.java [hunk 1]
```diff
@@ -149,21 +150,26 @@ static void safeClose(Channel ch) {
...
     private SessionProtocol protocol;
     private HttpObjectEncoder responseEncoder;
 
     private int unfinishedRequests;
     private boolean isReading;
     private boolean handledLastRequest;
 
-    HttpServerHandler(ServerConfig config, SessionProtocol protocol) {
+    HttpServerHandler(
+            ServerConfig config, GracefulShutdownSupport gracefulShutdownSupport, SessionProtocol protocol) {
```
Comment (imasahiro) [inline core/src/main/java/com/linecorp/armeria/server/http/HttpServerHandler.java]: Nit. Please add linebreak after `gracefulShutdownSupport,`  for readability.

File: core/src/main/java/com/linecorp/armeria/server/http/HttpServerPipelineConfigurator.java [hunk 2]
```diff
@@ -72,20 +72,20 @@
...
             ServerConfig config, ServerPort port,
             DomainNameMapping<SslContext> sslContexts,
-            Optional<? extends ChannelHandler> gracefulShutdownHandler) {
+            GracefulShutdownSupport gracefulShutdownSupport) {
 
         this.config = requireNonNull(config, "config");
         this.port = requireNonNull(port, "port");
         this.sslContexts = sslContexts;
-        this.gracefulShutdownHandler = requireNonNull(gracefulShutdownHandler);
+        this.gracefulShutdownSupport = requireNonNull(gracefulShutdownSupport);
```
Comment (imasahiro) [inline core/src/main/java/com/linecorp/armeria/server/http/HttpServerPipelineConfigurator.java]: Add `"gracefulShutdownSupport"`
 `requireNonNull(gracefulShutdownSupport)` ->  `requireNonNull(gracefulShutdownSupport, "gracefulShutdownSupport")`

File: core/src/main/java/com/linecorp/armeria/server/GracefulShutdownSupport.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2015 LINE Corporation
```
Comment (imasahiro) [inline core/src/main/java/com/linecorp/armeria/server/GracefulShutdownSupport.java]: 2015 -> 2017

**Comment by anuraaga** [comment]
LGTM, thanks!

**Comment by kojilin** [comment]
LGTM.

**Comment by trustin** [comment]
Thanks for reviewing, everyone!

**Pull request merged**